providers/okta/application.go

```diff
@@ -0,0 +1,252 @@
+	Method string
+}
+
+type IdentifyResponse struct {
+	StateHandle                    string
+	CurrentAuthenticatorEnrollment struct {
+		Type  string
+		Value AuthenticatorEnrollment
+	}
+
+	Remediation struct {
+		Type  string
+		Value []Remediation
+	}
+}
+
+func DetermineUpgradePath(resp IdentifyResponse) (MultiFactorUpgradeMethod, bool) {
+	// Two iterations is fine, this is a small list
+	// Prefer Duo OIDC flow
+	for _, rem := range resp.Remediation.Value {
+		if rem.Type == "OIDC" {
+			return DuoFrameless{remediation: rem}, true
+		}
+	}
+
+	for _, rem := range resp.Remediation.Value {
+		// This is intentionally 'Name' as Type is missing for cases where this is true.
+		if rem.Name == "challenge-authenticator" && resp.CurrentAuthenticatorEnrollment.Value.Key == "duo" {
+			host := resp.CurrentAuthenticatorEnrollment.Value.ContextualData.Host
+			tok, err := ParseSignedToken(resp.CurrentAuthenticatorEnrollment.Value.ContextualData.SignedToken)
+			if err != nil {
+				break
+			}
+
+			return DuoIframe{Host: host, SignedToken: tok, CallbackURL: rem.Href, Method: rem.Method, StateHandle: resp.StateHandle}, true
+		}
+	}
+
+	return nil, false
+}
+
+type ApplicationSAMLSource string
+
+func (source ApplicationSAMLSource) URL() string {
+	return string(source)
+}
+
+func (source ApplicationSAMLSource) Identify(ctx context.Context, client *http.Client, prevURL *url.URL, username, password string, stateToken StateToken) (identifyResponse IdentifyResponse, err error) {
+	identifyVals := map[string]any{
+		"credentials": map[string]string{
+			"passcode": password,
+		},
+		"identifier":  username,
+		"stateHandle": stateToken,
+	}
+	identifyBytes, _ := json.Marshal(identifyVals)
+	identifyURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/identify"})
+	resp, err := client.Post(identifyURL.String(), "application/json", bytes.NewReader(identifyBytes))
+	if err != nil {
+		return
+	}
+
+	if resp.StatusCode == http.StatusForbidden {
+		err = errors.New("user does not have access to this application")
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&identifyResponse)
+	return
+}
+
+type IntrospectResponse struct {
+	Success *struct {
+		Name string
+		Href string
+	}
+}
+
+func (source ApplicationSAMLSource) Introspect(ctx context.Context, client *http.Client, prevURL *url.URL, stateToken StateToken) (introspectResponse IntrospectResponse, err error) {
+	// We need to run introspect to get our next link.
+	// Now use /idp/idx/identify to get the IDP link to follow.
+	introspectVals := map[string]any{"stateToken": stateToken}
+	introspectBytes, _ := json.Marshal(introspectVals)
+	introspectURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/introspect"})
+	resp, err := client.Post(introspectURL.String(), "application/json", bytes.NewReader(introspectBytes))
+	if err != nil {
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&introspectResponse)
+	return
+}
+
+func extractStateToken(r io.Reader) (StateToken, error) {
```

Comment:
This is extremely cursed code and is quite likely to break if Okta's front page ever changes, but there is no real way around this.

In order for us to interact with Okta like a browser might, which is required for the OIDC flow to work,  we need to extract the state token within the initial request to Okta. Unfortunately, this request is _within the JavaScript rendered by Okta_, as Okta does entirely javascript redirections even with javascript disabled.

Okta, man, please, don't do this.

Reply:
I'd add the above as a comment in the code.

Reply:
will add